eta/core/serial.py

```diff
@@ -812,6 +818,550 @@ def _validate_dict(cls, d):
+            big._backing_dir = "./" + self._ELE_ATTR
+            big.write_json(index_path)
+            big._backing_dir = full_backing_dir
+
+            etau.make_archive(rootdir, archive_path)
+
+    @classmethod
+    def from_archive(
+            cls, archive_path, backing_dir=None, delete_archive=False):
+        '''Loads a Big iterable from an archive created by `to_archive()`.
+
+        Args:
+            archive_path: the path to the archive to load
+            backing_dir: optional backing directory to use for the new Big
+                iterable. If provided, it must be empty or non-existent
+            delete_archive: whether to delete the archive after unpacking it.
+                By default, this is False
+
+        Returns:
+            a Big iterable
+        '''
+        with etau.TempDir() as tmp_dir:
+            name = os.path.splitext(os.path.basename(archive_path))[0]
+            rootdir = os.path.join(tmp_dir, name)
+            index_path = os.path.join(rootdir, "index.json")
+            tmp_backing_dir = os.path.join(rootdir, cls._ELE_ATTR)
+
+            etau.extract_archive(
+                archive_path, outdir=tmp_dir, delete_archive=delete_archive)
+            big = cls.from_json(index_path)
+            big._backing_dir = tmp_backing_dir
+            big.move(backing_dir=backing_dir)
+
+        return big
+
+    @classmethod
+    def from_paths(cls, paths, backing_dir=None):
+        '''Creates a Big iterable from a list of `_ELE_CLS` JSON files.
+
+        Args:
+            backing_dir: optional backing directory to use for the new Big
+                iterable. If provided, it must be empty or non-existent
+            paths: an iterable of paths to `_ELE_CLS` JSON files
+
+        Returns:
+            a Big iterable
+        '''
+        big = cls(backing_dir=backing_dir)
+        for path in paths:
+            big._add_by_path(path)
+        return big
+
+    @classmethod
+    def from_dir(cls, source_dir, backing_dir=None):
+        '''Creates a Big iterable from an unstructured directory of `_ELE_CLS`
+        JSON files on disk.
+
+        The source directory is traversed recursively.
+
+        Args:
+            backing_dir: optional backing directory to use for the new Big
+                iterable. If provided, it must be empty or non-existent
+            source_dir: the source directory from which to ingest elements
+
+        Returns:
+            a Big iterable
+        '''
+        paths = etau.multiglob(".json", root=source_dir + "/**/*")
+        return cls.from_paths(paths, backing_dir=backing_dir)
+
+    def _set_backing_dir(self, backing_dir):
+        if backing_dir is not None:
+            self._uses_temporary_storage = False
+            self._backing_dir = os.path.abspath(backing_dir)
+        else:
+            self._uses_temporary_storage = True
+            self._backing_dir = etau.make_temp_dir()
+
+    @staticmethod
+    def _make_uuid():
```

Comment:
I don't think `BigMixin` should define these private methods- how the subclasses choose to store their elements on disk is class-specific. As evidence for this, `BigSet` uses `key` in these methods, but `BigContainer` uses `idx`

As I mentioned above, I think the right path forward is just to force `BigMixin` classes to implement `add_by_path(path)`

Reply:
Great point. Done!